src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -503,10 +500,23 @@ public void run() {
 						FastList<Partition> partitionsAsList = FastList.newList(partitionsToReset);
 						FastList<String> topics = partitionsAsList.collect(new PartitionToTopicFunction()).distinct();
 						kafkaTemplate.getConnectionFactory().refreshMetadata(topics);
-						Map<Partition, BrokerAddress> leaders = kafkaTemplate.getConnectionFactory().getLeaders(partitionsToReset);
-						synchronized (partitionsByBrokerMap) {
-							forEachKeyValue(leaders, new AddPartitionToBrokerProcedure());
-							partitionsByBrokerMap.notifyAll();
+
+						MutableSetMultimap<BrokerAddress, Partition> partitionsByBroker = UnifiedMap
+								.newMap(kafkaTemplate.getConnectionFactory().getLeaders(partitionsToReset)).flip();
+						for (Pair<BrokerAddress, RichIterable<Partition>> pair : partitionsByBroker
+								.keyMultiValuePairsView()) {
+							synchronized (fetchTasksByBroker) {
+								boolean addedSuccessfully = false;
+								FetchTask fetchTask = fetchTasksByBroker.get(pair.getOne());
+								if (fetchTask != null) {
+									addedSuccessfully = fetchTask.addListenedPartitionsIfActive(pair.getTwo());
```

Comment:
Yeah... There is no the problem with the `resetLeaders` in `catch` above.
You check the result of `addListenedPartitionsIfActive` here and start a new fresh task replacing the old one.

Well, I'm giving up. So distributed logic breaks my mind :smile: 

Maybe tomorrow from the fresh morning...